src/types/unsolicited_response.rs

```diff
@@ -0,0 +1,309 @@
+    /// that reports a change in a server or mailbox annotation.
+    Metadata {
+        /// Mailbox name for which annotations were changed.
+        mailbox: String,
+        /// List of annotations that were changed.
+        metadata_entries: Vec<String>,
+    },
+
+    /// An unsolicited [`VANISHED` response](https://tools.ietf.org/html/rfc7162#section-3.2.10)
+    /// that reports a sequence-set of `UID`s that have been expunged from the mailbox.
+    ///
+    /// The `VANISHED` response is similar to the `EXPUNGE` response and can be sent wherever
+    /// an `EXPUNGE` response can be sent. It can only be sent by the server if the client
+    /// has enabled [`QRESYNC`](https://tools.ietf.org/html/rfc7162).
+    ///
+    /// The `VANISHED` response has two forms, one with the `EARLIER` tag which is used to
+    /// respond to a `UID FETCH` or `SELECT/EXAMINE` command, and one without an `EARLIER`
+    /// tag, which is used to announce removals within an already selected mailbox.
+    ///
+    /// If using `QRESYNC`, the client can fetch new, updated and deleted `UID`s in a
+    /// single round trip by including the `(CHANGEDSINCE <MODSEQ> VANISHED)`
+    /// modifier to the `UID SEARCH` command, as described in
+    /// [RFC7162](https://tools.ietf.org/html/rfc7162#section-3.1.4). For example
+    /// `UID FETCH 1:* (UID FLAGS) (CHANGEDSINCE 1234 VANISHED)` would return `FETCH`
+    /// results for all `UID`s added or modified since `MODSEQ` `1234`. Deleted `UID`s
+    /// will be present as a `VANISHED` response in the `Session::unsolicited_responses`
+    /// channel.
+    Vanished {
+        /// Whether the `EARLIER` tag was set on the response
+        earlier: bool,
+        /// The list of `UID`s which have been removed
+        uids: Vec<std::ops::RangeInclusive<u32>>,
+    },
+
+    /// An unsolicited [`FLAGS` response](https://tools.ietf.org/html/rfc3501#section-7.2.6) that
+    /// identifies the flags (at a minimum, the system-defined flags) that are applicable in the
+    /// mailbox. Flags other than the system flags can also exist, depending on server
+    /// implementation.
+    ///
+    /// See [`Flag`] for details.
+    // TODO: the spec doesn't seem to say anything about when these may be received as unsolicited?
+    Flags(Vec<Flag<'static>>),
+
+    /// An unsolicited `OK` response.
+    ///
+    /// The `OK` response may have an optional `ResponseCode` that provides additional
+    /// information, per [RFC3501](https://tools.ietf.org/html/rfc3501#section-7.1.1).
+    Ok {
+        /// Optional response code.
+        code: Option<ResponseCode>,
+        /// Information text that may be presented to the user.
+        information: Option<String>,
+    },
+
+    /// An unsolicited `BYE` response.
+    ///
+    /// The `BYE` response may have an optional `ResponseCode` that provides additional
+    /// information, per [RFC3501](https://tools.ietf.org/html/rfc3501#section-7.1.5).
+    Bye {
```

Comment:
If we receive one of these, does that mean the connection is about to be terminated? Should we handle these directly in the library instead and turn them into an error?

Reply:
It does mean the server is about to close the connection, but the RFC also says the client [SHOULD continue to read responses off the wire](https://tools.ietf.org/html/rfc3501#section-7.1.5). So as is, this will turn into an unsolicited response, and then when the server will eventually hang up on us and the client will get an io Error.

I think the 'right' thing to do is to handle the `Bye` and then continue to read responses, which may be of the kind we're looking for or not, and then just let the server hang up on us. The client will get an error in both cases, but if we keep reading then the client may get a response between the `Bye` and the hangup. If we return the `Bye` as an error, then the client will probably terminate the connection and maybe lose those responses?

FWIW, I can go either way and would be happy for suggestions. As is, the IDLE client will get a callback with a `Bye` and then they can do whatever they want with it, and then they will eventually get an io Error if they choose to continue the connection. Since clients should be tolerant to sudden io errors anyway I don't think this will cause a problem in that way, but am not 100% sure if it will be a problem for other situations where we get an unsolicited `Bye` response while issuing a `SELECT` or something. Should we keep parsing after seeing the `Bye` (and maybe get a solicited response?) or return an Error to the user immediately?

Reply:
I think you've convinced me — we should surface this as an unsolicited BYE and let the client choose how to deal with it :+1: